ocaml/xapi/xapi_host.ml

```diff
@@ -2382,3 +2382,72 @@ let set_multipathing ~__context ~host ~value =
   Db.Host.add_to_other_config ~__context ~self:host ~key:"multipathing"
     ~value:(string_of_bool value) ;
   Xapi_host_helpers.Configuration.set_multipathing value
+
+let host_sched_gran_of_string s =
+  match String.lowercase_ascii s with
+  | "core" ->
+      `core
+  | "cpu" ->
+      `cpu
+  | "socket" ->
+      `socket
+  | _ ->
+      raise
+        (Api_errors.Server_error
+           ("Expected 'core','cpu', 'socket', got " ^ s, []))
+
+let host_sched_gran_to_string = function
+  | `core ->
+      "core"
+  | `cpu ->
+      "cpu"
+  | `socket ->
+      "socket"
+
+let set_sched_gran ~__context ~host ~sched =
+  if Helpers.get_localhost ~__context <> host then
+    failwith "Forwarded to the wrong host" ;
+  if not !Xapi_globs.allow_host_sched_gran_modification then
+    raise
+      Api_errors.(
+        Server_error (operation_not_allowed, ["Disabled by xapi.conf"])) ;
+  let arg = Printf.sprintf "sched-gran=%s" (host_sched_gran_to_string sched) in
+  let args = ["--set-xen"; arg] in
+  try
+    let _ = Helpers.call_script !Xapi_globs.xen_cmdline_script args in
+    ()
+  with e ->
+    error "Failed to update sched-gran: %s" (Printexc.to_string e) ;
+    raise
+      Api_errors.(
+        Server_error (internal_error, ["Failed to update sched-gran"]))
+
+let get_sched_gran ~__context ~host =
+  if Helpers.get_localhost ~__context <> host then
+    failwith "Forwarded to the wrong host" ;
+  let args = ["--get-xen"; "sched-gran"] in
+  try
+    let ret = Helpers.call_script !Xapi_globs.xen_cmdline_script args in
+    match ret with
+    | "\n" ->
+        `cpu
+    | "sched-gran=cpu\n" ->
```

Comment:
Bonus points for using the of_string function, as this is yet another enum/string conversion.